File: p0/p62/UniquePaths.java
package leetcode.p0.p62;

import java.util.Arrays;

public class UniquePaths {
    /**
     * DP
     * Time complexity: O(mn)
     * Space complexity: O(m)
     */
    public int uniquePaths(int m, int n) {
        int[] dp = new int[m];
        Arrays.fill(dp, 1);
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                dp[j] = dp[j] + dp[j - 1];
            }
        }
        return dp[m - 1];
    }
}

File: p0/p62/unique_paths.py
class Solution:
    """
    Feb 15, 2025 15:45
    Time Complexity: O(mn)
    Space Complexity: O(mn)
    """
    def uniquePaths(self, m: int, n: int) -> int:
        dp = []
        for i in range(m):
            dp.append([0] * n)

        for i in range(m):
            dp[i][0] = 1

        for i in range(n):
            dp[0][i] = 1

        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1]

        return dp[-1][-1]
